File: javascript/0015_3Sum.js
// https://leetcode.com/problems/3sum/
// O(n^2) time complexity
// O(log(n))
function threeSum(nums) {
  function twoSum(idx) {
    let [left, right] = [idx + 1, nums.length - 1];
    while (left < right) {
      const sum = nums[idx] + nums[left] + nums[right];
      if (sum > 0) {
        right -= 1;
      } else if (sum < 0) {
        left += 1;
      } else {
        output.push([nums[idx], nums[left++], nums[right--]]);
        while (left < right && nums[left - 1] === nums[left]) {
          left += 1;
        }
      }
    }
  }

  nums.sort((a, b) => a - b);
  const output = [];
  for (let i = 0; i < nums.length && nums[i] <= 0; i++) {
    if (i === 0 || nums[i - 1] !== nums[i]) {
      twoSum(i);
    }
  }
  return output;
}

File: leetcode/python/0015_3sum.py
# https://leetcode.com/problems/3sum/
# O(n^2) time complexity
# O(n) space complexity for worst case python sort implementation
def three_sum(nums):
    def two_sum(array, index, output):
        left, right = index + 1, len(array) - 1
        while left < right:
            total = array[index] + array[left] + array[right]
            if total < 0:
                left += 1
            elif total > 0:
                right -= 1
            else:
                output.append([array[index], array[left], array[right]])
                left += 1
                right -= 1
                while left < right and array[left] == array[left - 1]:
                    left += 1

    triplets = []
    nums.sort()
    for i, _ in enumerate(nums):
        if nums[i] > 0:
            break
        if i == 0 or nums[i] != nums[i - 1]:
            two_sum(nums, i, triplets)
    return triplets
